File: regions/system/library/go/bb-ai-client/http_client.go
package bbaiclient

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// HTTPClient は HTTP 経由で AI ゲートウェイと通信する AiClient の実装。
type HTTPClient struct {
	baseURL    string
	apiKey     string
	httpClient *http.Client
}

// HTTPClientConfig は HTTPClient の設定を保持する。
type HTTPClientConfig struct {
	BaseURL string
	APIKey  string
	Timeout time.Duration
}

// NewHTTPClient は HTTPClient を生成する。
// Timeout が 0 の場合は 30 秒がデフォルトとなる。
func NewHTTPClient(cfg HTTPClientConfig) *HTTPClient {
	timeout := cfg.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	return &HTTPClient{
		baseURL:    cfg.BaseURL,
		apiKey:     cfg.APIKey,
		httpClient: &http.Client{Timeout: timeout},
	}
}

// Complete は AI ゲートウェイの /v1/complete エンドポイントを呼び出す。
func (c *HTTPClient) Complete(ctx context.Context, req CompleteRequest) (*CompleteResponse, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/v1/complete", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+c.apiKey)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: do request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &APIError{StatusCode: resp.StatusCode}
	}

	var result CompleteResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("bbaiclient: decode response: %w", err)
	}
	return &result, nil
}

// Embed は AI ゲートウェイの /v1/embed エンドポイントを呼び出す。
func (c *HTTPClient) Embed(ctx context.Context, req EmbedRequest) (*EmbedResponse, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/v1/embed", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+c.apiKey)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: do request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &APIError{StatusCode: resp.StatusCode}
	}

	var result EmbedResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("bbaiclient: decode response: %w", err)
	}
	return &result, nil
}

// ListModels は AI ゲートウェイの /v1/models エンドポイントを呼び出す。
func (c *HTTPClient) ListModels(ctx context.Context) ([]ModelInfo, error) {
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, c.baseURL+"/v1/models", nil)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: create request: %w", err)
	}
	if c.apiKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+c.apiKey)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("bbaiclient: do request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &APIError{StatusCode: resp.StatusCode}
	}

	var result []ModelInfo
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("bbaiclient: decode response: %w", err)
	}
	return result, nil
}

File: regions/system/library/dart/bb-ai-client/lib/src/http_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import 'client.dart';
import 'types.dart';

// HTTP AI クライアント: AI ゲートウェイと HTTP 通信する実装
class HttpAiClient implements AiClient {
  final String baseUrl;
  final String apiKey;
  final Duration timeout;
  final http.Client _httpClient;

  HttpAiClient({
    required this.baseUrl,
    this.apiKey = '',
    // デフォルトタイムアウトは 30 秒
    this.timeout = const Duration(seconds: 30),
    http.Client? httpClient,
  }) : _httpClient = httpClient ?? http.Client();

  // AI ゲートウェイの /v1/complete エンドポイントを呼び出す
  @override
  Future<CompleteResponse> complete(CompleteRequest req) async {
    final raw = await _post('/v1/complete', req.toJson());
    return CompleteResponse.fromJson(raw);
  }

  // AI ゲートウェイの /v1/embed エンドポイントを呼び出す
  @override
  Future<EmbedResponse> embed(EmbedRequest req) async {
    final raw = await _post('/v1/embed', req.toJson());
    return EmbedResponse.fromJson(raw);
  }

  // AI ゲートウェイの /v1/models エンドポイントを呼び出す
  @override
  Future<List<ModelInfo>> listModels() async {
    final uri = Uri.parse('${baseUrl.replaceAll(RegExp(r'/$'), '')}/v1/models');
    final headers = <String, String>{};
    if (apiKey.isNotEmpty) {
      headers['Authorization'] = 'Bearer $apiKey';
    }
    final response = await _httpClient.get(uri, headers: headers).timeout(timeout);
    if (response.statusCode != 200) {
      throw AiClientError(
        'API error: ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
    final list = json.decode(response.body) as List;
    return list.map((e) => ModelInfo.fromJson(e as Map<String, dynamic>)).toList();
  }

  // POST リクエストを送信してレスポンス JSON を返す
  Future<Map<String, dynamic>> _post(String path, Map<String, dynamic> body) async {
    final uri = Uri.parse('${baseUrl.replaceAll(RegExp(r'/$'), '')}$path');
    final headers = <String, String>{'Content-Type': 'application/json'};
    if (apiKey.isNotEmpty) {
      headers['Authorization'] = 'Bearer $apiKey';
    }
    final response = await _httpClient
        .post(uri, headers: headers, body: json.encode(body))
        .timeout(timeout);
    if (response.statusCode != 200) {
      throw AiClientError(
        'API error: ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
    return json.decode(response.body) as Map<String, dynamic>;
  }
}
